### PR TITLE
Record: OrthoInit + Int6 MLP3x + BigramHash + SmearGate (val_bpb: 1.1539)

### PR DESCRIPTION
**Pull request by unnir**
# OrthoInit + Int6 MLP3x + BigramHash + SmearGate

## Approach

Six orthogonal improvements stacked on the baseline 9-layer, 512-dim GPT:

### 1. Orthogonal + muP-scaled Weight Initialization
- All large weight matrices initialized with orthogonal init (gain=1.0)
- Output projections (attn.proj, mlp.proj) scaled by `1/sqrt(2 * num_layers)` following muP
- Accelerates early convergence — the model starts closer to a well-conditioned point, giving Muon a head start

### 2. Int6 Mixed Quantization + zstd-22
- Per-row int6 quantization ([-32,31]) on MLP and attention weight matrices
- FP16 passthrough for tied embeddings and last 2 layers' Key projections (quantization-sensitive)
- zstd level 22 compression (better ratio than zlib-9 on int6 data)

### 3. 3x MLP Expansion
- MLP hidden dimension 1536 (3x model_dim), up from baseline 1024 (2x)
- Budget freed by int6 quantization pays for the extra parameters

### 4. Tuned Optimizer Hyperparameters
- `matrix_lr=0.02, scalar_lr=0.02, tied_embed_lr=0.03` (halved from baseline)
- `muon_momentum=0.99` with warmup from 0.92 over 1500 steps
- `warmdown_iters=3000`, `grad_clip_norm=0.3`
- `AdamW` with `weight_decay=0.01` for embedding/scalar params
- `beta1=0.9, beta2=0.95`

### 5. SmearGate + Bigram Hash Embedding
- SmearGate: learned gate blending each token's embedding with the previous token's (~512 params)
- Bigram Hash: 4096-bucket hash table (dim=128, projected to 512) injecting token-pair info

### 6. Training + Evaluation Setup
- `train_seq_len=2048, train_batch_tokens=786432`
- Sliding window evaluation with stride=64 at 2048-token windows

## Configuration

```bash
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

## Key Metrics

- Training: 7201 steps in 600s (83.33ms/step)
- Model params: 22,368,841
- Pre-quant: `val_bpb: 1.1696`
- Int6+zstd roundtrip: `val_bpb: 1.1748`
- **Sliding window (stride=64): `val_bpb: 1.1539`**
- Artifact: 15,162,375 bytes (under 16MB by 837,625 bytes)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: OrthoInit + Int6 MLP3x + BigramHash + SmearGate (val_bpb: 1.1539)

**BPB:** 1.1539 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `af51720b0d3c`, file `records/track_10min_16mb/2026-03-19_OrthoInit_Int6_MLP3x_BigramHash_SmearGate/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.04s, dim=512, layers=9, vocab=1024, code=63056 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.04s, dim=512, layers=9, vocab=1024, code=63056 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
